### PR TITLE
8284220: TypeMirror#toString omits enclosing class names after JDK-8281238

### PR DESCRIPTION
**Pull request by cushon**
Please consider this fix for https://bugs.openjdk.java.net/browse/JDK-8284220, which updates `ClassType#toString` to include enclosing class names when type annotations are present.

For example, `java.util.@A Map.Entry<?,?>` is now printed as `java.util.@A Map.Entry<?,?>`, instead of `java.util.@A Entry<?,?>` (note the missing `Map`).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 reviews required, with at least 1 reviewer)

### Issue
 * [JDK-8284220](https://bugs.openjdk.java.net/browse/JDK-8284220): TypeMirror#toString omits enclosing class names after JDK-8281238

### Reviewers
 * [Joe Darcy](https://openjdk.java.net/census#darcy) (@jddarcy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8084/head:pull/8084` \
`$ git checkout pull/8084`

Update a local copy of the PR: \
`$ git checkout pull/8084` \
`$ git pull https://git.openjdk.java.net/jdk pull/8084/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8084`

View PR using the GUI difftool: \
`$ git pr show -t 8084`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8084.diff">https://git.openjdk.java.net/jdk/pull/8084.diff</a>

</details>

### CODE REVIEW

File: test/langtools/tools/javac/tree/NestedTypeToString.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
```
Comment (jddarcy) [inline test/langtools/tools/javac/tree/NestedTypeToString.java:24]: The fix for JDK-8281238 required several existing tests to be updated, so it did not have a dedicated test written for it. For this follow-on fix, if the test is going to use an annotation processor, which is quite reasonable, I suggest:
* placing the test somewhere under test/langtools/tools/javac/processing/model
*  subclassing JavacTestingAbstractProcessor
*  using string-valued annotations to hold the expected result and checking them in the process method

There are other tests using processor following this model under test/langtools/tools/javac/processing/model.

Comment (cushon) [reply test/langtools/tools/javac/tree/NestedTypeToString.java]: Done, thanks for taking a look!

Comment (cushon) [reply test/langtools/tools/javac/tree/NestedTypeToString.java]: @jddarcy is there any other feedback I can address?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java [hunk 2]
```diff
@@ -1043,6 +1043,14 @@ public String toString() {
                         buf.append(tsym.packge());
                         buf.append(".");
                     }
+                    ListBuffer<Name> names = new ListBuffer<>();
```
Comment (jddarcy) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java:1046]: There may be a more idiomataic way to javac to construct this string, but offhand I don't know what it is.

File: test/langtools/tools/javac/processing/model/type/AnnotatedTypeToString/AnnotatedTypeToString.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (jddarcy) [inline test/langtools/tools/javac/processing/model/type/AnnotatedTypeToString/AnnotatedTypeToString.java:1]: From personal experience with similar tests, they can look obscure when revisting them months in the future. I suggest putting a sentence or two description of what the test is trying to do. Also, if all the types can fit in one file, that can be easier to understand.

**Review by jddarcy — APPROVED** [review]

**Review by jddarcy — APPROVED** [review]
Looks fine.

**Pull request closed without merging**